File: apps/backend/internal/interfaces/ws/protocol.go
package ws

// MessageType identifies the role of a WebSocket message.
type MessageType string

const (
	// Client → Server
	TypeCommand MessageType = "command"

	// Server → Client
	TypeOutput MessageType = "output"
	TypeDone   MessageType = "done"
	TypeError  MessageType = "error"
	TypeStatus MessageType = "status"
)

// ClientMessage is a message sent from the browser to the server.
type ClientMessage struct {
	Type    MessageType `json:"type"`
	Payload string      `json:"payload"`
}

// ServerMessage is a message sent from the server to the browser.
type ServerMessage struct {
	Type    MessageType `json:"type"`
	Payload string      `json:"payload"`
}

File: apps/frontend/src/lib/types/protocol.ts
export type MessageType = 'command' | 'output' | 'done' | 'error' | 'status' | 'cancel';

/** Message sent from client to server. */
export interface ClientMessage {
  type: MessageType;
  request_id: string;
  command?: string;
  timestamp: string;
}

/** Message sent from server to client. */
export interface ServerMessage {
  type: MessageType;
  request_id?: string;
  content?: string;
  timestamp: string;
}
